ui/src/treegrid.tsx

```diff
@@ -0,0 +1,253 @@
+
+export type TreeGridId = string | number;
+
+export type TreeGridRowData = {
+  [key: string]: string | number | JSX.Element;
+};
+
+export type TreeGridItem = {
+  children?: TreeGridId[];
+  data: TreeGridRowData;
+};
+
+export type TreeGridData = {
+  [key: TreeGridId]: TreeGridItem;
+};
+
+export type TreeGridColumnData = {
+  key: string;
+  width: string | number;
+  title?: string | JSX.Element;
+};
+
+export type TreeGridProps = {
+  columns: TreeGridColumnData[];
+  data: TreeGridData;
+  prefixElements?: (id: TreeGridId, data: TreeGridRowData) => JSX.Element[];
+  loadChildren?: (id: TreeGridId) => Promise<string>;
+};
+
+export function TreeGrid(props: TreeGridProps) {
+  const [state, updateState] = useImmer<TreeGridState>(
+    new Map<TreeGridId, TreeGridItemState>()
+  );
+
+  const toggleExpanded = useCallback((id: TreeGridId) => {
+    updateState((draft) => {
+      const itemState = draft.get(id) || {
+        status: Status.Collapsed,
+      };
+      switch (itemState.status) {
+        case Status.Loading:
+          return;
+        case Status.Expanded:
+          itemState.status = Status.Collapsed;
+          break;
+        default:
+          if (!props.loadChildren || !!itemState.loaded) {
+            itemState.status = Status.Expanded;
+          } else {
+            itemState.status = Status.Loading;
+            props
+              .loadChildren(id)
+              .then(() => {
+                updateState((draft) => {
+                  draft.set(id, {
+                    status: Status.Expanded,
+                    loaded: true,
+                  });
+                });
+              })
+              .catch((error) => {
+                updateState((draft) => {
+                  draft.set(id, {
+                    status: Status.Failed,
+                    errorMessage: error,
+                  });
+                });
+              });
+          }
+      }
+      draft.set(id, itemState);
+    });
+  }, []);
+
+  return (
+    <>
+      <table
+        style={{
+          tableLayout: "fixed",
+        }}
+      >
+        <thead>
+          <tr>
+            {props.columns.map((col, i) => (
+              <td
+                key={i}
+                style={{
+                  ...(col.width && {
+                    maxWidth: col.width,
+                    width: col.width,
+                    minWidth: col.width,
+                  }),
+                }}
+              >
+                <Typography variant="h6">{col.title}</Typography>
+              </td>
+            ))}
+          </tr>
+        </thead>
+      </table>
+      <div style={{ overflowY: "auto", display: "block" }}>
+        <table
+          style={{
+            tableLayout: "fixed",
+          }}
+        >
+          <tbody>
+            {renderChildren(props, state, toggleExpanded, "root", 0, false)}
+          </tbody>
+        </table>
+      </div>
+    </>
+  );
+}
+
+function renderChildren(
```

Comment:
That's the design I started with but it mostly just shuffled things around. Since the root node isn't rendered, the code to render children needs to be shared between the tree and potential child components anyway, and the actual rendering is pretty contained inside the results.push call.